File: Projetos/aeds2/tp01/c/TP01Q02 - Palíndromo em C/palindrome.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define MAX_LENGTH 255

void readln(char string[]) {
  fgets(string, MAX_LENGTH, stdin);

  const unsigned len = strlen(string);

  if (len > 0 && string[len - 1] == '\n') {
    string[len - 1] = '\0';
  }
}

bool isPalindrome(char string[]) {
  bool palindrome = true;
  int start = 0, end = (int) strlen(string) - 1;

  while (start < end) {
    if (string[start] != string[end]) {
      palindrome = false;
      end = 0;
    }

    start++;
    end--;
  }

  return palindrome;
}

int main() {
  char line[MAX_LENGTH];
  readln(line);

  while (strcmp("FIM", line) != 0) {
    printf("%s\n", isPalindrome(line) ? "SIM" : "NAO");;

    readln(line);
  }

  return 0;
}

File: Projetos/aeds2/tp01/java/TP01Q01 - Palíndromo/Palindrome.java
import java.util.Scanner;

class Palindrome {
    static Scanner scanner = new Scanner(System.in);

    static boolean isPalindrome(String string) {
        boolean palindrome = true;
        int start = 0, end = string.length() - 1;

        while (start < end) {
            if (string.charAt(start) != string.charAt(end)) {
                palindrome = false;
                end = 0;
            }

            start++;
            end--;
        }

        return palindrome;
    }

    static boolean isEOF(String line) {
        boolean eof = true;
        String flag = "FIM";

        if (flag.length() == line.length()) {
            for (int i = 0; i < line.length(); i++) {
                if (line.charAt(i) != flag.charAt(i)) {
                    eof = false;
                    i = flag.length();
                }
            }
        } else {
            eof = false;
        }

        return eof;
    }

    public static void main(String[] args) {
        String input = scanner.nextLine();

        while (!isEOF(input)) {
            System.out.println(isPalindrome(input) ? "SIM" : "NAO");

            input = scanner.nextLine();
        }
    }
}
